src/modelfree/configs/multi/train.py

```diff
@@ -15,6 +17,90 @@
+                          "final_model",
+        "best_hyper_10e6": "hyper_finetune_defense/20190923_222110-bfd8c5be86334254aff81681a1aab"
+                           "fd2/train_rl-d28a8df8c1c8116007a96f0a1324e4d5_62_batch_size=2048,learn"
+                           "ing_rate=6.1888e-05,ent_coef=0.0079898,nminibatches=1,noptepochs_"
+                           "2019-09-23_22-21-12atw_xvi3/data/baselines/20190923_222140-default"
+                           "-batch_size=2048-learning_rate=6.188842944950313e-05-rl_args="
+                           "{'ent_coef': 0.007989770218617331, 'nminibatches': 1, 'noptepochs': 6}"
+                           "-seed=261/final_model",
+        "best_seed_20e6_3e6params": "ysnp_finetune_20_dual/20190923_225840-bcfe315e866548418648a2"
+                                    "9ec51d7f91/train_rl-148c15a21ac330cec3747094c5ac5972_4_seed="
+                                    "4_2019-09-23_22-58-40469tpn1v/data/baselines/20190923_225913-"
+                                    "default-seed=4/final_model"
+    },
+    "dual_train":
+        {
+            "best_hyper_3e6": "hyper_finetune_dual_defense/20190919_230454-2a21dcece0bb420783b083a"
+                              "8c9bca393/train_rl-a01b394e1af3e513adbc965f57f28105_38_batch_size=8"
+                              "192,learning_rate=0.00019521,ent_coef=0.0012618,nminibatches=32,nop"
+                              "tepoch_2019-09-19_23-04-5585r1bib4/data/baselines/20190919_230522-d"
+                              "efault-batch_size=8192-learning_rate=0.00019520504186694475-rl_args"
+                              "={'ent_coef': 0.001261805277214847, 'nminibatches': 32,"
+                              " 'noptepochs': 1}-seed=601/final_model",
+            "best_hyper_10e6": "hyper_finetune_defense/20190923_222140-cc0e22f56cad42a9bfef263edb1"
+                               "b9747/train_rl-fb0cc754cf65767d85e5d39aa31552e8_78_batch_size=4096"
+                               ",learning_rate=3.1928e-05,ent_coef=0.0058516,nminibatches=32,nopt"
+                               "epoch_2019-09-23_22-21-42mquddv27/data/baselines/20190923_222201"
+                               "-default-batch_size=4096-learning_rate=3.192828794681526e-05-rl_"
+                               "args={'ent_coef': 0.00585162756249535, 'nminibatches': 32, "
+                               "'noptepochs': 2}-seed=444/final_model",
+            "best_seed_20e6_3e6params": "ysnp_finetune_20_dual/20190923_225942-dc84f4936b8147a0af6"
+                                        "6fcc4bc80ae2a/train_rl-8fd3be16c7723c9dceaec25a74aac1a8_1"
+                                        "_seed=1_2019-09-23_22-59-42yxnox92y/data/baselines/201909"
+                                        "23_230005-default-seed=1/final_model"
+
+        }
+}
+
+YSNP_ADVERSARY = "paper/20190429_011349/train_rl-7086bd7945d8a380b53e797f3932c739_10_env_name:" \
```

Comment:
Okay. I've been avoiding building on the existing code to get adversary paths because it's so nested and difficult to reason about how to actually extend it beyond the specific use case we currently have, but I can try to dive into some of today.

Reply:
OK, it might be easiest to discuss in person how this works and possible ways of refactoring.

I sympathise with not wanting to dig into deeply nested code, but if the existing code is too awkward to work with I think we need to fix that code. Adding hard-coded paths is just adding more tech debt. It won't scale to when we run this experiment across all the environments. And if we ever improve our adversary training code, we'd have to change the paths all again, etc.